src/main/java/ca/uhn/fhir/jpa/starter/CustomConsentService.java

```diff
@@ -0,0 +1,173 @@
+      case "POST":
+        resourceName = theRequestDetails.getResourceName();
+        proceed = isResourceValid(resourceName, patientId, theRequestDetails.getResource());
+        break;
+      case "PUT":
+        resourceName = theRequestDetails.getResourceName();
+        existingResource = getResourceFromDB(theRequestDetails.getRequestPath());
+        if (isResourceValid(resourceName, patientId, existingResource)) {
+          proceed = isResourceValid(resourceName, patientId, theRequestDetails.getResource());
+        }
+        break;
+      case "PATCH":
+        resourceName = theRequestDetails.getResourceName();
+        existingResource = getResourceFromDB(theRequestDetails.getRequestPath());
+        if (isResourceValid(resourceName, patientId, existingResource)) {
+          // As Patch request body doesn't contain any Resource we need to handle it
+          // differently
+          proceed = isPatchRequestBodyValid(resourceName,
+              new String(theRequestDetails.getRequestContentsIfLoaded()));
+        }
+        break;
+      default:
+        proceed = true;
+        break;
+    }
+    return proceed ? ConsentOutcome.PROCEED : ConsentOutcome.REJECT;
+  }
+
+  @Override
+  public ConsentOutcome canSeeResource(RequestDetails theRequestDetails, IBaseResource theResource,
+      IConsentContextServices theContextServices) {
+    /*
+     * Returning authorized if there is no Authorization header present or if requested resource is
+     * present in Patient Compartment For both this cases all the consent logic is in authorization
+     * intercepter rules
+     */
+    if (ObjectUtils.isEmpty(theRequestDetails.getHeader("Authorization"))
+        || oAuth2Helper.canBeInPatientCompartment(theRequestDetails.getResourceName())) {
+      return ConsentOutcome.AUTHORIZED;
+    }
+    String patientId = getPatientFromToken(theRequestDetails);
+    String resourceName = theResource.getClass().getSimpleName();
+    return isResourceValid(resourceName, patientId, theResource) ? ConsentOutcome.PROCEED
+        : ConsentOutcome.REJECT;
+  }
+
+  private IBaseResource getResourceFromDB(String requestedPath) {
+    String[] requestDetail = requestedPath.split("/");
+    return daoRegistry.getResourceDao(requestDetail[0]).read(new IdType(requestDetail[1]));
+  }
+
+  private boolean isResourceValid(String resourceName, String patientId,
+      IBaseResource theResource) {
+    if (patientId != null) {
+      String patientRef = "Patient/" + patientId;
+      switch (resourceName) {
+        case "Task":
+          return isReferanceValid((Resource) theResource, patientRef, "for");
+        default:
+          return false;
+      }
+    }
+    return true;
+  }
+
+  private boolean isReferanceValid(Resource theResource, String patientRef,
+      String refPropertyName) {
+    try {
+      Reference ref = (Reference) theResource.getNamedProperty(refPropertyName).getValues().get(0);
```

Comment:
Is it valid to get the first element of the array? i.e. Is it possible there would be more than 1?

Reply:
It should always come as 0th value but still, I will iterate over a loop.

Reply:
@shubhamparikh927 My question is more about the validity of having more than 1 `for` reference? Is that valid? What would we do with such a resource to ensure that the data is not leaked to the wrong patient?

Reply:
@hankwallace As of my knowledge, I didn't found multiple for reference in any Task resource I referred to their site also.

Reply:
The `for` attribute is single cardinality.